testing/src/main/java/org/opennms/nephron/testing/flowgen/SyntheticFlowSource.java

```diff
@@ -0,0 +1,124 @@
+
+import net.jqwik.api.RandomGenerator;
+
+/**
+ * A source of synthetically generated flows.
+ *
+ * Flow generation can be parameterized by a {@link SourceConfig} instance. Flows are generated deterministically
+ * base on the supplied seed value for a random number generator.
+ */
+public class SyntheticFlowSource extends UnboundedSource<FlowDocument, FlowReader.CheckpointMark> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SyntheticFlowSource.class);
+
+    /**
+     * Creates a transformation that reads from a synthetic source.
+     */
+    public static PTransform<PBegin, PCollection<FlowDocument>> readFromSyntheticSource(SourceConfig sourceConfig) {
+        return Read.from(new SyntheticFlowSource(sourceConfig));
+    }
+
+    private final SourceConfig sourceConfig;
+
+    public SyntheticFlowSource(SourceConfig sourceConfig) {
+        this.sourceConfig = sourceConfig;
+    }
+
+    @Override
+    public List<? extends UnboundedSource<FlowDocument, FlowReader.CheckpointMark>> split(int desiredNumSplits, PipelineOptions options) throws Exception {
+        return sourceConfig.split().stream()
+                .map(sc -> new SyntheticFlowSource(sc))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public UnboundedReader<FlowDocument> createReader(PipelineOptions options, FlowReader.@Nullable CheckpointMark checkpointMark) throws IOException {
+        RandomGenerator<FlowDocuments.FlowData> flowData = FlowDocuments.getFlowDataArbitrary(sourceConfig.flowConfig).generator(1000);
+        Optional<Instant> previous;
+        long startIdx;
+        Random random;
+        if (checkpointMark == null) {
+            LOG.trace("creating initial unbounded reader for {}", options);
+            previous = Optional.empty();
+            startIdx = sourceConfig.idxOffset;
+            random = new Random(sourceConfig.seed);
+        } else {
+            LOG.trace("resuming unbounded reader from {}", checkpointMark);
+            previous = Optional.of(checkpointMark.previous);
+            startIdx = checkpointMark.index;
+            random = checkpointMark.random;
+        }
+        Limiter limiter = Limiter.of(sourceConfig.flowsPerSecond);
+        return new FlowReader(
+                this,
+                (rnd, idx) -> FlowDocuments.getFlowDocument(sourceConfig.flowConfig, idx, flowData.next(rnd).value()),
+                sourceConfig.timestampPolicyFactory.create(previous),
+                sourceConfig.idxInc,
+                sourceConfig.maxIdx,
+                startIdx,
+                random,
+                limiter
+        );
+
+    }
+
+    @Override
+    public Coder<FlowReader.CheckpointMark> getCheckpointMarkCoder() {
+        return FlowReader.CheckpointMark.CHECKPOINT_CODER;
+    }
+
+    @Override
+    public Coder<FlowDocument> getOutputCoder() {
+        // TODO why do we need this? the FlowDocumentProbufCoder was registered with the pipeline
```

Comment:
Is this something that we have to check before merging?

Reply:
IMO this has not to be investigated. I just noted that I do not understand the necessity of returning the output encoder. I can remove that comment if you like.

Reply:
This is fine, but please change the comment an attach your handle like so:
```suggestion
        // TODO swachter: why do we need this? the FlowDocumentProbufCoder was registered with the pipeline
```